### ISSUE TITLE
Failed to match boolean value

### ISSUE THREAD

**Issue by Bab64** [nikku/feelin#71]
Maybe I misunderstood something, but the two behaviors (trivial Boolean logic) below are weird?

- [https://nikku.github.io/feel-playground/?e=true&c=%7B%0A++%22%3F%22+%3A+false%0A%7D%0A&t=unaryTests&st=true](https://nikku.github.io/feel-playground/?e=true&c=%7B%0A++%22%3F%22+%3A+false%0A%7D%0A&t=unaryTests&st=true)
- [https://nikku.github.io/feel-playground/?e=false&c=%7B%0A++%22%3F%22+%3A+false%0A%7D%0A&t=unaryTests&st=true](https://nikku.github.io/feel-playground/?e=false&c=%7B%0A++%22%3F%22+%3A+false%0A%7D%0A&t=unaryTests&st=true)

**Comment by nikku** [comment]
Sounds like a bug indeed. Evaluation should match by value, i.e. [this](https://nikku.github.io/feel-playground/?e=%22120%22&c=%7B%0A++%22%3F%22+%3A+%22120%22%0A%7D%0A&t=unaryTests&st=true).

**Comment by Skaiir** [comment]
Just so that I understand the issue properly. 

What about this scenario, what should the unary test return?

https://nikku.github.io/feel-playground/?e=3+%3C+1&c=%7B%0A++%22%3F%22+%3A+false%0A%7D%0A&t=unaryTests&st=true

Do we ONLY match by value in the very specific case where both the value and test are a simple boolean?

**Comment by Bab64** [comment]
I believe it's already fixed up by nikku? The initial problem was comparisons with zero (**not** the case here). Again, fixed, I guess.

**Comment by nikku** [comment]
> I believe it's already fixed up by nikku? The initial problem was comparisons with zero (not the case here). Again, fixed, I guess.

No, this is not fixed (yet). `?` denotes the current value, it is matched against a unary test.

If we enter a test that is always truthy `3 > 1 = true` then we match whatever `?` is against the result of that test, i.e. `true` if `? = true`, else `false`.

**Comment by Skaiir** [comment]
Looking into our own documentation:

> A unary-tests expression returns true if one of the following conditions is fulfilled:
> 
>  -  !! The expression evaluates to true when the input value is applied to it.
>  -  The expression evaluates to a list, and the input value is equal to at least one of the values in that list.
>  -  !! The expression evaluates to a value, and the input value is equal to that value.
>  -  The expression is equal to - (a dash).

 
The issue here is that for booleans, the two scenarios highlighted via !! are ambiguous. A small improvement would be to always compare boolean values to their test by value, but then it leads to other errors:

```
?: true*, test: < 5
true < 5 becomes 1 < 5 becomes true**
true* === true** becomes true ✔️ 

?: false*, test: < 5
false < 5 becomes 0 < 5 becomes true**
false* === true** becomes false ❌ 
```

So I guess the "bug" that exists within the codebase is that we do not discriminate between when the expression itself is driven by the test value vs when it isn't.

An example, if I understand it correctly, is that we expect `false, 0 > 1 => true` but `false, ? > 1 => false` @nikku

**Comment by nikku** [comment]
@Skaiir I'm not sure I fully understand what you mean.

What would help is if you'd file a PR that contributes [test cases](https://github.com/nikku/feelin/blob/main/test/interpreter-spec.js#L8849) to show which cases currently conflict. 

Even better we could raise this with the [DMN TCK](https://github.com/dmn-tck/tck) to get clarity across all implementors.

**Comment by nikku** [comment]
Related discussion in `feel-scala`: https://github.com/camunda/feel-scala/issues/864#issuecomment-2188837091.

TLDR: We want to assess our current handling of such cases.

**Comment by nikku** [comment]
This library gets the following wrong still, cf. https://github.com/camunda/feel-scala/issues/864#issuecomment-2188837091, spec excerpt:

![image](https://github.com/user-attachments/assets/2fbf5d3e-d9b2-4f0f-bd24-a7347f2fb117)

```
{ input: null, test: 'null > 20', expectedOutput: true }
{ input: null, test: 'not(null > 20)', expectedOutput: false }
```

**Comment by nikku** [comment]
Fixed via [v4.4.0](https://github.com/nikku/feelin/releases/tag/v4.4.0).

**Comment by nikku** [comment]
You can verify the fixed behavior in the [playground](https://nikku.github.io/feel-playground/).
